### PR TITLE
#130: added more filters to outline view and changed sorting

### PR DESCRIPTION
**Pull request by jpilgrim**
See #130, note that #126 is still pending

- [x] filter for static member
- [x] filter for public member
- [x] filter for local member
- [x] sort takes meta types into account
- [x] improved coloring
- [x] improved labels (better signatures, e.g., type args, typing strategy)
- [x] build (branch GH-130)

### CODE REVIEW

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend [hunk 1]
```diff
@@ -195,6 +200,9 @@ class StyledTextCalculationHelper {
 	 */
 	def dispatch StyledString dispatchGetStyledText(TSetter tsetter) {
 		var styledText = getLabelProvider.getSuperStyledText(tsetter)
+		if (styledText === null) return unknown;
+
+		styledText.setStyle(0, styledText.length, N4JSStylers.FIELD_OR_VAR_STYLER)
 		if (tsetter.fpar !== null && tsetter.fpar !== null) {
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend]: Duplicate condition `tsetter.fpar !== null`

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend [hunk 2]
```diff
@@ -216,7 +224,12 @@ class StyledTextCalculationHelper {
 	 */
 	def dispatch StyledString dispatchGetStyledText(TField tfield) {
 		var styledText = getLabelProvider.getSuperStyledText(tfield);
-		if (tfield.typeRef !== null && styledText !== null) {
+		if (styledText === null) return unknown;
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend]: Similar to fields, getters and setters can also be optional. We need to add `?` to getters and setters if needed as well

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend [hunk 3]
```diff
@@ -216,7 +224,12 @@ class StyledTextCalculationHelper {
...
+		styledText.setStyle(0, styledText.length, N4JSStylers.FIELD_OR_VAR_STYLER)
+		if (tfield.isOptional) {
+			styledText.append("?");
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend]: The style/color for the field name is **different** from that of the question mark.
`styledText.append("?", N4JSStylers.FIELD_OR_VAR_STYLER);` instead?

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/outline/MetaTypeAwareComparator.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import org.eclipse.xtext.ui.editor.outline.impl.EObjectNode;
+import org.eclipse.xtext.ui.editor.outline.impl.OutlineFilterAndSorter.IComparator;
+import org.eclipse.xtext.ui.editor.preferences.IPreferenceStoreAccess;
+
+import com.google.inject.Inject;
+
+/**
+ *
+ */
+public class MetaTypeAwareComparator implements IComparator {
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/outline/MetaTypeAwareComparator.java]: What is the purpose of this type-aware sorting? As a user, when I toggle "A -> Z" sort button, I intuitively expect the list to be sorted alphabetically. But the order presented seems to be random now...

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend (hunk 1 above)
Comment (jpilgrim) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend]: done

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend (hunk 2 above)
Comment (jpilgrim) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend]: done

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend (hunk 3 above)
Comment (jpilgrim) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend]: I did that on purpose: the "?" is not part of the name. Makes it slightly better to read (I tested that myself ;-) ). Actually I introduce the field colors after finding myself looking for the fields in longer lists.

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/outline/MetaTypeAwareComparator.java (hunk 4 above)
Comment (jpilgrim) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/outline/MetaTypeAwareComparator.java]: I know. The problem is: I'm not quite sure how to add several sorters in the outline view. The API is a little bit strange. In the long run, we can change that. For the time being I wanted the things to be sorted as I did it. It is a pragmatic solution.

Oh, and the order is far from random! It is simply metatype based now, i.e. fields are sorted before methods.

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend (hunk 3 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/labeling/helper/StyledTextCalculationHelper.xtend]: OK it make sense :)

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/outline/MetaTypeAwareComparator.java (hunk 4 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/outline/MetaTypeAwareComparator.java]: OK we can improve this in the future.

**Pull request merged**